## _Домашнее задание 1_
Скачать данные по ссылке <BR>
https://www.kaggle.com/datasets/ionaskel/laptop-prices <BR>
Считать данные с помощью pandas <BR>
Вывести на экран первые 5 строк<BR>
Посмотреть на описание признаков и на их содержание
    

In [7]:
import pandas as pd
import numpy as np

In [8]:
df = pd.read_csv('./kc_house_data.csv', sep=',' , encoding='windows-1251')
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0          1180.0              0      1955             0    98178  47.5112   
1          2170.0            400      1951          1991    98125  47.7210   
2           770.0              0      1933             0    98028  47.7379   
3          1050.0            910      1965             0    98136  47.5208   
4          1680.0              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608      1530.0              0      2009             0    98103  47.6993   
21609      2310.0              0      2014             0    98146  47.5107   
21610      1020.0              0      2009             0    98144  47.5944   
21611      1600.0              0      2004             0    98027  47.5345   
21612      1020.0              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [58]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

* id - Уникальный ID для каждого дома
* date - Дата продажи дома
* price - Стоимость продажи дома
* bedrooms - Кол-во спален
* bathrooms - Кол-во ванных комнат (0.5 - туалет без душа)
* sqft_living - Кв. метры жилые
* sqft_lot - Кв. метры общие
* floors - Кол-во этажей
* waterfront - Есть набержная или нет
* view - Значение от 0 до 4 насколько хороший вид
* condition - Значение от 1 до 5 насколько хорошее состояние
* grade - Значение от 1 до 13, где 1-3 плохая конструкция здания и дизайн, 7 - средний уровень конструкции и дизайна, 11-13 - высокое качество конструкции и дизайна
* sqft_above - Кв. метры дома, которые находятся выше земли
* sqft_basement - Кв. метры дома, которые находятся ниже земли
* yr_built - Год постройки дома
* yr_renovated - Год ремонта дома
* zipcode - Индекс
* lat - Широта
* long - Долгота
* sqft_living15 - Кв. метры жилой площади у 15 соседей
* sqft_lot15 - Кв. метры общей площади у 15 соседей

# _Домашнее задание 2_
Проведите первичный анализ данных
1. Изучите типы данных
2. Найдите количество пропущенных ячеек в данных
3. Посчитайте основные статистики по всем признакам и
поизучайте их
4. Пишите выводы

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [6]:
df.describe(include='object')

date
count             21613
unique              372
top     20140623T000000
freq                142

In [7]:
df.describe(include='float64')

price     bathrooms        floors    sqft_above           lat  \
count  2.161300e+04  21613.000000  21613.000000  21611.000000  21613.000000   
mean   5.400881e+05      2.114757      1.494309   1788.396095     47.560053   
std    3.671272e+05      0.770163      0.539989    828.128162      0.138564   
min    7.500000e+04      0.000000      1.000000    290.000000     47.155900   
25%    3.219500e+05      1.750000      1.000000   1190.000000     47.471000   
50%    4.500000e+05      2.250000      1.500000   1560.000000     47.571800   
75%    6.450000e+05      2.500000      2.000000   2210.000000     47.678000   
max    7.700000e+06      8.000000      3.500000   9410.000000     47.777600   

               long  
count  21613.000000  
mean    -122.213896  
std        0.140828  
min     -122.519000  
25%     -122.328000  
50%     -122.230000  
75%     -122.125000  
max     -121.315000

# Выводы:
1. Получены типы данных df.info().
2. В таблице нет пропущенных ячеек, все колонки содержат 21613 ненулевых значений.
3. Статистика по всем признакам выше, 372 уникальные записи из 21613, частота появления самого частотного значения - 142,<bp>
самое частотное значение (мода) - дата 20140623T000000

# _Домашнее задание 3_
Ответьте на несколько вопросов<br>
3.1 В каком диапазоне изменяются стоимости недвижимости?<br>
3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?<br>
3.3 Как много домов с разными этажами в данных?<br>
3.4 Насколько хорошие состояния у домов в данных?<br>
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [8]:
df['price'].min()

75000.0

In [9]:
df['price'].max()

7700000.0

* 3.1 Стоимость недвижимости изменяется в диапазоне от 75 до 7700 тыс.усл.единиц

In [18]:
df_living=df['sqft_living'].sum()
df_living

44952873

In [19]:
df_all=df['sqft_lot'].sum()
df_all

326506890

In [22]:
result_3_2=df_living/df_all
result_3_2

0.1376781757959227

* 3.2  долю жилой площади от всей площади по всем домам - 14%

In [24]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

* 3.3 Количесвто домов с разными этажами в данных
1.0  | 10680 <br>
2.0  |   8241<br>
1.5  |   1910<br>
3.0  |    613<br>
2.5  |    161<br>
3.5  |      8<br>

In [30]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

* 3.4 Насколько хорошие состояния у домов в данных?
3  |  14031<br>
4  |   5679<br>
5  |   1701<br>
2  |    172<br>
1  |     30<br>

In [34]:
df.iloc[[0, 21612]][['yr_built']]

yr_built
0          1955
21612      2008

* 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных? <br>
Первый дом по списку построен в 1955 году, второй - в 2008 году

# _Домашнее задание 4_
Ответьте на несколько вопросов<br>
4.1 Сколько в среднем стоят дома, у которых 2 спальни?<br>
4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?<br>
4.3 Как много домов коснулся ремонт?<br>
4.4 Насколько в среднем стоимость домов с оценкой grade<br>
домов выше 10 отличается от стоимости домов с оценкой grade<br>
меньше 4?

* 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [38]:
df_2_bathrooms=df[(df['bathrooms']==2)][['price']].mean()
df_2_bathrooms

price    457889.718653
dtype: float64

* 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [44]:
df_sqft_lot=df[(df['price']>600000)][['sqft_lot']].mean()
df_sqft_lot

sqft_lot    20442.524776
dtype: float64

* 4.3 Как много домов коснулся ремонт?

In [53]:
df_renovated=df[(df['yr_renovated']>0)][['yr_renovated']].shape[0]
df_renovated

914

* 4.4 Насколько в среднем стоимость домов с оценкой grade <br>
домов выше 10 отличается от стоимости домов с оценкой grade
меньше 4?

In [60]:
df_grade1=df[(df['grade']>10)][['price']].mean()
df_grade1

price    1.678635e+06
dtype: float64

In [61]:
df_grade2=df[(df['grade']<4)][['price']].mean()
df_grade2

price    189750.0
dtype: float64

В среднем стоимость домов с оценкой:<br>
    * grade домов выше 10 - 1 678 635<br>
    * grade домов меньше 4 -  189 750

# _Домашнее задание 5_
## 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя
ванными и с подвалом. Сколько вариантов есть у клиента?
## 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с
видом на набережную, в очень хорошем состоянии и год
постройки не меньше 1980 года. В какой ценовом диапазоне
будут дома?
## 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью
до 150000. Какая оценка по состоянию у таких домов в
среднем?

In [68]:
df_vibor_5_1=df[(df['waterfront']==1)&(df['bathrooms']>=3)&(df['sqft_basement']>0)].shape[0]
df_vibor_5_1

41

In [10]:
df_vibor_5_2=df[((df['waterfront']==1)|(df['view']==4))&(df['condition']==5)&(~(df['yr_built']<1980))]
df_vibor_5_2.sort_values('price')

id             date      price  bedrooms  bathrooms  \
14682  7278100515  20140821T000000  1295000.0         2       2.50   
18991  7856400300  20140702T000000  1411600.0         2       2.50   
18992  7856400300  20150322T000000  1505000.0         2       2.50   
17312  3585901025  20140613T000000  1735000.0         3       2.50   
11257  3761100045  20140618T000000  3000000.0         4       4.25   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
14682         2910     19449     2.0           1     4  ...      9   
18991         3180      9400     2.0           0     4  ...     10   
18992         3180      9400     2.0           0     4  ...     10   
17312         4310     32093     1.5           0     4  ...     10   
11257         4850     12445     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
14682      1940.0            970      1985             0    98177  47.7729   
18991      2610.0            570      1985             0    98006  47.5617   
18992      2610.0            570      1985             0    98006  47.5617   
17312      2510.0           1800      1982             0    98177  47.7624   
11257      3850.0           1000      1989             0    98034  47.7011   

          long  sqft_living15  sqft_lot15  
14682 -122.393           2540       23598  
18991 -122.158           3760        9450  
18992 -122.158           3760        9450  
17312 -122.380           3810       26400  
11257 -122.244           3350       12210  

[5 rows x 21 columns]

In [13]:
df_vibor_5_3=df[(df['sqft_basement']==0)&(df['floors']==2)&(df['price']<150000)]
df_vibor_5_3

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2589           980      5000     2.0           0     0  ...      7   
10105         1230      1561     2.0           0     0  ...      7   
15610         1270      8991     2.0           0     0  ...      7   
18332         1200      7000     2.0           0     0  ...      7   
18939          980      2490     2.0           0     0  ...      6   
18988         1200      2002     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2589        980.0              0      1922          2003    98014  47.7076   
10105      1230.0              0      1983             0    98042  47.3768   
15610      1270.0              0      1981             0    98042  47.3563   
18332      1200.0              0      1908             0    98116  47.5883   
18939       980.0              0      1969             0    98092  47.3170   
18988      1200.0              0      1966             0    98055  47.4659   

          long  sqft_living15  sqft_lot15  
2589  -121.359           1040        5000  
10105 -122.149           1660        2243  
15610 -122.149           1270        8993  
18332 -122.384           3290        6000  
18939 -122.182            980        3154  
18988 -122.189           1270        1848  

[6 rows x 21 columns]

In [15]:
df_vibor_5_3['condition'].mean()

2.8333333333333335